File: relay/src/bin/conformance.rs
//! Cross-language conformance runner for relay protocol v1.
//!
//! Reads the vector payload from stdin (see `protocol/conformance.py`) and emits
//! one JSON object mapping vector name -> {canonical, signature}. The Python
//! driver diffs this against the Worker's TypeScript output; any disagreement in
//! canonical form or HMAC fails CI.
//!
//! This exists so the three independent protocol implementations cannot drift
//! silently — before the monorepo they only shared a hand-copied fixture.

use std::collections::BTreeMap;
use std::io::Read;

use codex_https_relay::protocol::signing::{
    build_canonical_request, sign_relay_request, RelaySigningInput,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
struct Payload {
    secret: String,
    nonce: String,
    vectors: Vec<Vector>,
}

#[derive(Debug, Deserialize)]
struct Vector {
    name: String,
    key_id: String,
    timestamp: i64,
    method: String,
    target: String,
    headers: Vec<[String; 2]>,
    body_utf8: String,
}

#[derive(Debug, Serialize)]
struct Outcome {
    canonical: String,
    signature: String,
}

fn main() {
    let mut raw = String::new();
    std::io::stdin()
        .read_to_string(&mut raw)
        .expect("read stdin payload");
    let payload: Payload = serde_json::from_str(&raw).expect("parse payload");

    let mut results: BTreeMap<String, Outcome> = BTreeMap::new();
    for vector in &payload.vectors {
        let body = vector.body_utf8.as_bytes();
        let input = RelaySigningInput {
            version: 1,
            key_id: &vector.key_id,
            timestamp: vector.timestamp,
            nonce: &payload.nonce,
            method: &vector.method,
            target: &vector.target,
            headers: &vector.headers,
            body,
        };

        let canonical = build_canonical_request(&input)
            .unwrap_or_else(|error| panic!("vector {}: canonical failed: {error:?}", vector.name));
        let signature = sign_relay_request(&input, payload.secret.as_bytes())
            .unwrap_or_else(|error| panic!("vector {}: signing failed: {error:?}", vector.name));

        results.insert(
            vector.name.clone(),
            Outcome {
                canonical,
                signature,
            },
        );
    }

    println!(
        "{}",
        serde_json::to_string(&results).expect("serialize results")
    );
}

File: claude-ingress/scripts/conformance.ts
/**
 * Cross-language conformance runner for relay protocol v1.
 *
 * Reads the vector payload from stdin (see `protocol/conformance.py`) and emits
 * one JSON object mapping vector name -> {canonical, signature}. The Python
 * driver diffs this against the Rust relay's output; any disagreement in
 * canonical form or HMAC fails CI.
 *
 * Runs on plain Node (not workerd) so the gate stays cheap. `crypto.subtle` and
 * `btoa`/`atob` are globals in Node 22, matching the Worker runtime's API.
 */
import { buildCanonicalRequest } from "../src/relay/protocol";
import { sha256Base64Url, signRelayRequest } from "../src/relay/signing";

type Vector = {
  name: string;
  /**
   * Protocol generation this vector is signed under.
   *
   * Carried per vector rather than per payload so one run drives both
   * generations. A runner that ignored this field would agree with the others on
   * v1 and silently never exercise v2.
   */
  version: number;
  key_id: string;
  timestamp: number;
  method: string;
  target: string;
  headers: Array<[string, string]>;
  body_utf8: string;
};

type Payload = {
  secret: string;
  nonce: string;
  vectors: Vector[];
};

async function readStdin(): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    chunks.push(Buffer.from(chunk));
  }
  return Buffer.concat(chunks).toString("utf8");
}

const payload = JSON.parse(await readStdin()) as Payload;
const results: Record<string, { canonical: string; signature: string }> = {};

for (const vector of payload.vectors) {
  const body = new TextEncoder().encode(vector.body_utf8);
  const input = {
    version: vector.version,
    keyId: vector.key_id,
    timestamp: vector.timestamp,
    nonce: payload.nonce,
    method: vector.method,
    target: vector.target,
    headers: vector.headers,
    body,
  };

  const bodySha256 = await sha256Base64Url(body);
  results[vector.name] = {
    canonical: buildCanonicalRequest(input, bodySha256),
    signature: await signRelayRequest(input, payload.secret),
  };
}

process.stdout.write(`${JSON.stringify(results)}\n`);
